peps/pep-0728.rst

```diff
@@ -0,0 +1,614 @@
+the type of ``__extra__``::
+
+    def f(movie: Movie, key: str) -> None:
+        reveal_type(movie["name"])  # Revealed type is 'str'
+        reveal_type(movie["year"])  # Revealed type is 'int'
+
+
+Interaction with Totality
+---------------------------
+
+It is an error to use ``Required[]`` or ``NotRequired[]`` with the special
+``__extra__`` field. ``total=False`` and ``total=True`` have no effect on
+``__extra__`` itself.
+
+The extra fields are non-required, regardless of the totality of the TypedDict.
+Operations that are available to ``NotRequired`` fields should also be available
+to the extra fields::
+
+    def f(movie: Movie) -> None:
+        del movie["name"]  # Not OK
+        del movie["year"]  # OK
+
+Interaction with ``Unpack``
+---------------------------
+
+For type checking purposes, ``Unpack[TypedDict]`` with extra fields should be
+treated as its equivalent in regular parameters, and the existing rules for
+function parameters still apply::
+
+    class Movie(TypedDict):
+        name: str
+        __extra__: int
+    
+    def f(**kwargs: Unpack[Movie]) -> None: ...
+
+    # Should be equivalent to
+    def f(*, name: str, **kwargs: int) -> None: ...
+
+Inheritance
+-----------
+
+``__extra__`` is inherited the same way as a regular ``key: value_type``
+field. The same rules from :pep:`PEP 589 <589#inheritance>` apply.
+
+Additionally, we need to reinterpret the following rule to define how
+``__extra__`` interacts with it:
+
+    * Changing a field type of a parent TypedDict class in a subclass is not allowed.
+
+First, it is not allowed to change the type of ``__extra__`` in a subclass::
+
+    class Parent(TypedDict):
+        __extra__: int | None
+    
+    class Child(Parent):
+        __extra__: int  # Not OK. Like any other TypedDict field, __extra__'s type cannot be changed
+
+Second, ``__extra__`` effectively defines the type of any unnamed fields
+accepted to the TypedDict and marks them as non-required. Thus, the above
+restriction applies to any additional fields defined in a subclass. For each
+field added in a subclass, all of the following conditions should apply:
+
+- The field is non-required
+
+- The field's value type is consistent with the type of ``__extra__``
+
+- The type of ``__extra__`` is consistent with the field's value type
+
+For example::
+
+    class MovieBase(TypedDict):
+        name: str
+        __extra__: int | None
+    
+    class AdaptedMovie(MovieBase):  # Not OK. 'bool' is not consistent with 'int | None'
+        adapted_from_novel: bool
+ 
+    class MovieRequiredYear(MovieBase):  # Not OK. Required key 'x' is not known to 'Parent'
+        year: int | None
+
+    class MovieNotRequiredYear(MovieBase):  # Not OK. 'int | None' is not consistent with 'int'
+        year: NotRequired[int]
+
+    class MovieWithYear(MovieBase):  # OK
+        x: NotRequired[int | None]
+
+As a side-effect, annotating ``__extra__`` with :class:`typing.Never` for a
```

Comment:
It may be worth making this change much louder than as a side-effect.

This is a key feature people have wanted independent of the more general question of typing extra fields, and in fact type checkers have at times used `@final` for this but there's a very reasonable view that `@final` is inappropriate.

I brought up the problem of exact TypedDict in [this typing thread](https://discuss.python.org/t/do-we-want-an-exact-typeddict-if-so-how-final-extras-never/44418/4) and @erictraut pointed me at PEP 728, but it took me a little while to find this aside.

Reply:
Updated the PEP to extend on this topic.